diff -r -wBu "Python38\\Lib\\distutils/_msvccompiler.py" "Python3\\Lib\\distutils/_msvccompiler.py"
--- "Python38\\Lib\\distutils/_msvccompiler.py"	2020-07-13 16:52:40.000000000 +0800
+++ "Python3\\Lib\\distutils/_msvccompiler.py"	2020-07-14 15:29:38.491012500 +0800
@@ -161,6 +161,27 @@
             return fn
     return exe
 
+def _find_llvm():
+    try:
+        key = winreg.OpenKeyEx(
+            winreg.HKEY_LOCAL_MACHINE,
+            r"SOFTWARE\LLVM\LLVM",
+            access=winreg.KEY_READ|winreg.KEY_WOW64_32KEY
+        )
+    except OSError:
+        log.debug("LLVM is not registered")
+        return None
+
+    with key:
+        for i in count():
+            try:
+                v, llvm_dir, vt = winreg.EnumValue(key, i)
+            except OSError:
+                break
+            if vt == winreg.REG_SZ and os.path.isdir(llvm_dir):
+               return llvm_dir
+    return None
+
 # A map keyed by get_platform() return values to values accepted by
 # 'vcvarsall.bat'. Always cross-compile from x86 to work with the
 # lighter-weight MSVC installs that do not include native 64-bit tools.
@@ -171,6 +192,13 @@
     'win-arm64' : 'x86_arm64'
 }
 
+PLAT_TO_LLVM_TARGETS = {
+    'win32': 'i686-pc-windows-msvc',
+    'win-amd64': 'x86_64-pc-windows-msvc',
+    'win-arm32' : 'arm-pc-windows-msvc',
+    'win-arm64': 'aarch64-pc-windows-msvc',
+}
+
 class MSVCCompiler(CCompiler) :
     """Concrete class that implements an interface to Microsoft Visual C++,
        as defined by the CCompiler abstract class."""
@@ -202,11 +230,12 @@
     exe_extension = '.exe'
 
 
-    def __init__(self, verbose=0, dry_run=0, force=0):
+    def __init__(self, verbose=0, dry_run=0, force=0, use_clang_cl=False):
         CCompiler.__init__ (self, verbose, dry_run, force)
         # target platform (.plat_name is consistent with 'bdist')
         self.plat_name = None
         self.initialized = False
+        self.use_clang_cl = use_clang_cl
 
     def initialize(self, plat_name=None):
         # multi-init means we would need to check platform same each time...
@@ -228,7 +257,14 @@
 
         self._paths = vc_env.get('path', '')
         paths = self._paths.split(os.pathsep)
-        self.cc = _find_exe("cl.exe", paths)
+        if self.use_clang_cl:
+            llvm_dir = _find_llvm()
+            if llvm_dir:
+                self.cc = os.path.join(llvm_dir, 'bin', 'clang-cl.exe')
+            else:
+                self.cc = _find_exe("clang-cl.exe", paths)
+        else:
+            self.cc = _find_exe('cl.exe', paths)
         self.linker = _find_exe("link.exe", paths)
         self.lib = _find_exe("lib.exe", paths)
         self.rc = _find_exe("rc.exe", paths)   # resource compiler
@@ -263,6 +299,17 @@
             '/nologo', '/INCREMENTAL:NO', '/LTCG', '/DEBUG:FULL'
         ]
 
+        if self.use_clang_cl:
+            # Add target flags for clang
+            target_flag = '--target=' + PLAT_TO_LLVM_TARGETS[plat_name]
+            self.compile_options.append(target_flag)
+            self.compile_options_debug.append(target_flag)
+            # Remove whole program optimization and Link-time code generation flags
+            # to avoid warnings about unrecognized options
+            self.compile_options.remove('/GL')
+            ldflags.remove('/LTCG')
+            ldflags_debug.remove('/LTCG')
+
         self.ldflags_exe = [*ldflags, '/MANIFEST:EMBED,ID=1']
         self.ldflags_exe_debug = [*ldflags_debug, '/MANIFEST:EMBED,ID=1']
         self.ldflags_shared = [*ldflags, '/DLL', '/MANIFEST:EMBED,ID=2', '/MANIFESTUAC:NO']
@@ -541,3 +588,9 @@
         else:
             # Oops, didn't find it in *any* of 'dirs'
             return None
+
+class ClangMSVCCompiler(MSVCCompiler):
+    compiler_type = 'clang-cl'
+
+    def __init__(self, verbose=0, dry_run=0, force=0):
+        super().__init__(verbose, dry_run, force, True)
diff -r -wBu "Python38\\Lib\\distutils/ccompiler.py" "Python3\\Lib\\distutils/ccompiler.py"
--- "Python38\\Lib\\distutils/ccompiler.py"	2020-07-13 16:52:40.000000000 +0800
+++ "Python3\\Lib\\distutils/ccompiler.py"	2020-07-14 15:27:26.076494700 +0800
@@ -962,6 +962,8 @@
                                "standard UNIX-style compiler"),
                    'msvc':    ('_msvccompiler', 'MSVCCompiler',
                                "Microsoft Visual C++"),
+                   'clang-cl':('_msvccompiler', 'ClangMSVCCompiler',
+                               "clang-cl for Microsoft Visual C++"),
                    'cygwin':  ('cygwinccompiler', 'CygwinCCompiler',
                                "Cygwin port of GNU C Compiler for Win32"),
                    'mingw32': ('cygwinccompiler', 'Mingw32CCompiler',
